ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -323,24 +323,62 @@ let assert_host_has_iommu ~__context ~host =
   if List.assoc "iommu" chipset_info <> "true" then
     raise (Api_errors.Server_error (Api_errors.vm_requires_iommu, [Ref.string_of host]))
 
-let assert_gpus_available ~__context ~self ~host =
-  let vgpus = Db.VM.get_VGPUs ~__context ~self in
-  let reqd_groups =
-    List.map (fun self -> Db.VGPU.get_GPU_group ~__context ~self) vgpus in
+let pgpus_for_vgpu ~__context ~host ~vgpu =
   let is_pgpu_available pgpu vgpu =
     try Xapi_pgpu.assert_can_run_VGPU ~__context ~self:pgpu ~vgpu; true
     with _ -> false
   in
+  let group = Db.VGPU.get_GPU_group ~__context ~self:vgpu in
+  let pgpus = Db.GPU_group.get_PGPUs ~__context ~self:group in
+  let avail_pgpus =
+    List.filter
+      (fun pgpu -> is_pgpu_available pgpu vgpu)
+      pgpus
+  in
+  List.filter (fun self -> Db.PGPU.get_host ~__context ~self = host) avail_pgpus
+
+let assert_gpus_compatible ~__context ~self ~host =
+  let vgpus = Db.VM.get_VGPUs ~__context ~self in
+  let open Gpumon_interface in
+  let get_reasons = function
+    | Compatible -> []
+    | Incompatible reasons -> reasons
+  in
+  List.iter (fun vgpu ->
+      let open Xapi_vgpu_type in
+      let vgpu_impl = vgpu
+        |> (fun self -> Db.VGPU.get_type ~__context ~self)
+        |> (fun self -> Db.VGPU_type.get_implementation ~__context ~self)
+      in
+      match vgpu_impl with
+      | `passthrough | `gvt_g | `mxgpu -> ()
+      | `nvidia ->
+        let pgpus =
+          match Db.VGPU.get_scheduled_to_be_resident_on ~__context ~self:vgpu with
```

Comment:
I am not convinced about this `match`, I should probably have two different functions one used in the `pool_migrate` and one in `assert_can_migrate`. I made one function because there was a lot of boilerplate in common